modules/camera/camera_win.cpp

```diff
@@ -29,32 +29,382 @@
+	UINT32 count; // Number of elements in the array.
+	UINT32 selection; // Selected device, by array index.
+};
+
+class VideoBufferLock {
+public:
+	VideoBufferLock(IMFMediaBuffer *pBuffer) :
+			m_p2DBuffer(NULL), m_bLocked(FALSE) {
+		m_pBuffer = pBuffer;
+		m_pBuffer->AddRef();
+
+		// Query for the 2-D buffer interface. OK if this fails.
+		(void)m_pBuffer->QueryInterface(IID_PPV_ARGS(&m_p2DBuffer));
+	}
+
+	~VideoBufferLock() {
+		UnlockBuffer();
+		SafeRelease(&m_pBuffer);
+		SafeRelease(&m_p2DBuffer);
+	}
+
+	//-------------------------------------------------------------------
+	// LockBuffer
+	//
+	// Locks the buffer. Returns a pointer to scan line 0 and returns the stride.
+	//
+	// The caller must provide the default stride as an input parameter, in case
+	// the buffer does not expose IMF2DBuffer. You can calculate the default stride
+	// from the media type.
+	//-------------------------------------------------------------------
+
+	HRESULT LockBuffer(
+			LONG lDefaultStride, // Minimum stride (with no padding).
+			DWORD dwHeightInPixels, // Height of the image, in pixels.
+			BYTE **ppbScanLine0, // Receives a pointer to the start of scan line 0.
+			LONG *plStride // Receives the actual stride.
+	) {
+		HRESULT hr = S_OK;
+
+		// Use the 2-D version if available.
+		if (m_p2DBuffer) {
+			hr = m_p2DBuffer->Lock2D(ppbScanLine0, plStride);
+		} else {
+			// Use non-2D version.
+			BYTE *pData = NULL;
+
+			hr = m_pBuffer->Lock(&pData, NULL, NULL);
+			if (SUCCEEDED(hr)) {
+				*plStride = lDefaultStride;
+				if (lDefaultStride < 0) {
+					// Bottom-up orientation. Return a pointer to the start of the
+					// last row *in memory* which is the top row of the image.
+					*ppbScanLine0 = pData + abs(lDefaultStride) * (dwHeightInPixels - 1);
+				} else {
+					// Top-down orientation. Return a pointer to the start of the
+					// buffer.
+					*ppbScanLine0 = pData;
+				}
+			}
+		}
+
+		m_bLocked = (SUCCEEDED(hr));
+
+		return hr;
+	}
+
+	//-------------------------------------------------------------------
+	// UnlockBuffer
+	//
+	// Unlocks the buffer. Called automatically by the destructor.
+	//-------------------------------------------------------------------
+
+	void UnlockBuffer() {
+		if (m_bLocked) {
+			if (m_p2DBuffer) {
+				(void)m_p2DBuffer->Unlock2D();
+			} else {
+				(void)m_pBuffer->Unlock();
+			}
+			m_bLocked = FALSE;
+		}
+	}
+
 private:
+	IMFMediaBuffer *m_pBuffer;
+	IMF2DBuffer *m_p2DBuffer;
+
+	BOOL m_bLocked;
+};
+
+__forceinline BYTE Clip(int clr) {
+	return (BYTE)(clr < 0 ? 0 : (clr > 255 ? 255 : clr));
+}
+
+__forceinline RGBQUAD ConvertYCrCbToRGB(
+		int y,
+		int cr,
+		int cb) {
+	RGBQUAD rgbq;
+
+	int c = y - 16;
+	int d = cb - 128;
+	int e = cr - 128;
+
+	rgbq.rgbRed = Clip((298 * c + 409 * e + 128) >> 8);
+	rgbq.rgbGreen = Clip((298 * c - 100 * d - 208 * e + 128) >> 8);
+	rgbq.rgbBlue = Clip((298 * c + 516 * d + 128) >> 8);
+
+	return rgbq;
+}
+
+void TransformImage_YUY2(
+		BYTE *pDest,
```

Comment:
For the MacOS camera stuff I made a shader that can handle YUV so we didn't need to do conversions on the CPU side. Is it worth looking at this further and adding YUY2 support as well or does it make more sense to have this RGB conversion in? 

I believe there is a conversion library already linked in as a 3rd party module?

Reply:
can't seem to find the conversion library, but found some reference code to do it within MF itself. Curious why the provided sample files didn't use them. Going to implement them.